src/menu_init.c

```diff
@@ -0,0 +1,23 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   menu_init.c                                        :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: gwolf <gwolf@student.42vienna.com>         +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/12/30 16:34:39 by gwolf             #+#    #+#             */
+/*   Updated: 2024/01/06 16:12:10 by gwolf            ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "render_menu.h"
+
+void	ft_init_menu(t_menu *menu, uint8_t alpha, uint32_t colour,
+			uint32_t font_col)
+{
+	menu->inv_alpha = 255 - alpha;
+	menu->rb = (colour & 0xFF00FF) * alpha;
```

Comment:
How come we don't differentiate between red and blue? I read that you did for performance reasons, but why is that ok to handle them together? And why these two and not green and blue for example?

Reply:
It actually makes use of two simplifications:
- less variables
- less precision while division

A color is stored in a 4-byte int, where the four bytes represent alpha, red, green and blue or `ARGB`. For simple alpha blending the formula is 
ColorC = alpha{ColorA} * ColorA + (1 - alpha{ColorA} ) * ColorB. 
So we mult first color with alpha and second color with inverse of alpha. Then we add them together.

We need to multiply channel-wise, or we get funky results. This is often done by splitting the color in three variables, one for each channel (R,G,B), multiply with alpha and then stick together.
Here comes the first optimization: we only use two variables. 
Since we store the color in 4 bytes, we can make use of the remaining space. We isolate the R and B channels so `ARGB`  becomes `0R0B`. We can envision it as having two "zero points", from which we can multiply. This means R and B do not overlap, they don't interfere with each other while multiplying. For that we remove the G channel with `& 0xFF00FF`. The int now only stores R and B.  
We still need to handle G so we remove the others with `& 0x00FF00` which leaves us `00G0`.
Then we can safely multiply the channels with the inverse of alpha.
This is what happens here (the equivalent for the menu channels RB and G happens at menu_init):
```
bg_rb = (bg_color & 0xFF00FF) * menu.inv_alpha;
bg_g = (bg_color & 0x00FF00) * menu.inv_alpha;
```

We add the RB and G, each with its alpha value together. We save alpha in one byte, so it ranges from 0 - 255. This means we need to divide the resulting number by 255 (or 255.99 or whatever approximation is used) to get the actual channel value. (Alternative would be to save alpha as float in range 0.0 to 1.0 which uses 4 bytes). 
Here comes the second optimization: we bit shift 8 to the right (`>> 8`), which effectively divides by 2^8 which is 256. It won't give us the truest color blend, but a good enough approximation with speed gains. To ensure we're only left with the desired channels we `&` them again.
```
blend_rb = (((menu.rb + bg_rb) >> 8) & 0xFF00FF);
blend_g = (((menu.g + bg_g) >> 8) & 0x00FF00);
```

After that we only need to stick the two variables together which we can do with bitwise OR: `0R0B` | `00G0` = `0RGB` (we could encode the alpha in here, but we don't need it, so it is left out). This happens in the return.
`return (blend_rb | blend_g);`